### PR TITLE
feat: shooter

### PR DESCRIPTION
**Pull request by TiaSinghania**
Draft flywheel code - needs to be modified + tested on actual prototype shooter, once it's built. Could test on R2Tree2 but motor types would need to be changed.

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -78,4 +78,8 @@
     public static final int kDriverControllerPort = 0;
     public static final double kJoystickDeadzone = 0.11;
   }
+
+  public static final class ShooterConstants {
+    public static final int kShooterMotorPort = 9;
+  }
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/Constants.java:84]: should be moved above OIConstants

File: src/main/java/frc/robot/HardwareMap.java [hunk 2]
```diff
@@ -1,14 +1,39 @@
...
 
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+import com.ctre.phoenix.motorcontrol.can.VictorSPX;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
+import com.ctre.phoenix.motorcontrol.can.WPI_VictorSPX;
+import com.ctre.phoenix.sensors.CANCoder;
+import edu.wpi.first.wpilibj.Servo;
+import edu.wpi.first.wpilibj.SpeedControllerGroup;
+import edu.wpi.first.wpilibj.XboxController;
+import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/HardwareMap.java:11]: organize imports

File: src/main/java/frc/robot/HardwareMap.java [hunk 3]
```diff
@@ -1,14 +1,39 @@
...
+        }
+    }
+    
+    public class InputHardware {
+        public XboxController driveController;
+        public InputHardware() {
+            driveController = new XboxController(Constants.OIConstants.kDriverControllerPort);
+        }
+    }
+    ;
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/HardwareMap.java:35]: should be in separate PR

File: src/main/java/frc/robot/HardwareMap.java [hunk 4]
```diff
@@ -1,14 +1,39 @@
...
-/** File for storing the hardware of the robot. */
 public class HardwareMap {
+
+    public class ShooterHardware{
+        public  WPI_TalonFX flywheel;
+
+        public ShooterHardware(){
+            flywheel = new WPI_TalonFX(Constants.ShooterConstants.kShooterMotorPort);          
+        }
+    }
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/HardwareMap.java:27]: should be moved below Module Hardware

File: src/main/java/frc/robot/HardwareMap.java (hunk 3 above)
Comment (LennyAbbas) [inline src/main/java/frc/robot/HardwareMap.java:35]: delete

File: src/main/java/frc/robot/HardwareMap.java [hunk 5]
```diff
@@ -1,14 +1,39 @@
...
 import com.kauailabs.navx.frc.AHRS;
 import com.revrobotics.CANSparkMax;
 import com.revrobotics.CANSparkMaxLowLevel.MotorType;
 
 import frc.robot.Constants.SwerveConstants;
 import frc.robot.subsystems.SwerveModule;
 
-/** File for storing the hardware of the robot. */
 public class HardwareMap {
+
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/HardwareMap.java:20]: need doc

File: src/main/java/frc/robot/HardwareMap.java [hunk 6]
```diff
@@ -1,14 +1,39 @@
...
+import edu.wpi.first.wpilibj.XboxController;
+import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
 import com.kauailabs.navx.frc.AHRS;
 import com.revrobotics.CANSparkMax;
 import com.revrobotics.CANSparkMaxLowLevel.MotorType;
 
 import frc.robot.Constants.SwerveConstants;
 import frc.robot.subsystems.SwerveModule;
 
-/** File for storing the hardware of the robot. */
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/HardwareMap.java:10]: why remove?

File: src/main/java/frc/robot/HardwareMap.java [hunk 7]
```diff
@@ -72,10 +97,15 @@
 
   public SwerveModuleHardware swerveModuleHardware;
   public SwerveDrivetrainHardware swerveDrivetrainHardware;
+
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/HardwareMap.java:100]: remove line

File: src/main/java/frc/robot/HardwareMap.java [hunk 8]
```diff
@@ -72,10 +97,15 @@
...
 
   /** Creates a new {@link HardwareMap}. */
   public HardwareMap() {
     swerveModuleHardware = new SwerveModuleHardware();
     swerveDrivetrainHardware = new SwerveDrivetrainHardware();
+    shooterHardware = new ShooterHardware();
+    inputHardware = new InputHardware();
   }
-}
+}
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/HardwareMap.java:111]: what was changed here?

File: src/main/java/frc/robot/Main.java [hunk 9]
```diff
@@ -4,6 +4,8 @@
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/Main.java:4]: Main.java should not be changed

File: src/main/java/frc/robot/RobotContainer.java [hunk 10]
```diff
@@ -10,6 +10,11 @@
 import edu.wpi.first.wpilibj.GenericHID;
 import edu.wpi.first.wpilibj.XboxController;
 import edu.wpi.first.wpilibj.XboxController.Button;
+import frc.robot.HardwareMap.ShooterHardware;
+import frc.robot.commands.ShooterCommand;
+import frc.robot.subsystems.ShooterSubsystem;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.button.JoystickButton;
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/RobotContainer.java:17]: organize imports

File: src/main/java/frc/robot/RobotContainer.java [hunk 11]
```diff
@@ -19,6 +24,7 @@
 import frc.robot.commands.SwerveDriveCommand;
 import frc.robot.subsystems.SwerveDriveSubsystem;
 
+
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/RobotContainer.java:27]: bad

File: src/main/java/frc/robot/RobotContainer.java [hunk 12]
```diff
@@ -27,13 +33,18 @@
  * commands, and button mappings) should be declared here.
  */
 public class RobotContainer {
+
+  // The robot's subsystems and commands are defined here...
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/RobotContainer.java:37]: delete

File: src/main/java/frc/robot/RobotContainer.java [hunk 13]
```diff
@@ -27,13 +33,18 @@
...
   private HardwareMap m_hardwareMap = new HardwareMap();
+  private ShooterHardware ShooterHardware = m_hardwareMap.shooterHardware;
+  private ShooterSubsystem m_shooterSubsystem = new ShooterSubsystem(ShooterHardware);
+
+  private ShooterCommand m_ShooterCommand = new ShooterCommand(m_shooterSubsystem);
+  private XboxController m_driveController = m_hardwareMap.inputHardware.driveController;
   private SwerveDriveSubsystem m_swerveDriveSubsystem = new SwerveDriveSubsystem(
       m_hardwareMap.swerveDrivetrainHardware);
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/RobotContainer.java:45]: reorder

File: src/main/java/frc/robot/RobotContainer.java [hunk 14]
```diff
@@ -62,6 +73,9 @@ public RobotContainer() {
    * passing it to a {@link edu.wpi.first.wpilibj2.command.button.JoystickButton}.
    */
   private void configureButtonBindings() {
+
+    new JoystickButton(m_driveController, Button.kY.value)
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/RobotContainer.java:77]: need comment

File: src/main/java/frc/robot/commands/ShooterCommand.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.subsystems.ShooterSubsystem;
+
+public class ShooterCommand extends CommandBase {
+
+  private ShooterSubsystem m_shooterSubsystem; 
+  private PIDController m_pidController = new PIDController(0.7, 0, 0);
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/commands/ShooterCommand.java:15]: PID should go above, both member vars should be final

File: src/main/java/frc/robot/commands/ShooterCommand.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+  /** Creates a new ShooterCommand. */
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/commands/ShooterCommand.java:17]: need javadoc

File: src/main/java/frc/robot/commands/ShooterCommand.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+  public ShooterCommand(ShooterSubsystem shooterSubsystem) {
+    // Use addRequirements() here to declare subsystem dependencies.
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/commands/ShooterCommand.java:19]: remove

File: src/main/java/frc/robot/commands/ShooterCommand.java [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    m_shooterSubsystem = shooterSubsystem;
+    addRequirements(m_shooterSubsystem);
+  }
+
+  // Called when the command is initially scheduled.
+  @Override
+  public void initialize() {}
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/commands/ShooterCommand.java:26]: remove

File: src/main/java/frc/robot/commands/ShooterCommand.java [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+  // Called every time the scheduler runs while the command is scheduled.
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/commands/ShooterCommand.java:28]: remove

File: src/main/java/frc/robot/commands/ShooterCommand.java [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+  @Override
+  public void execute() { 
+    m_pidController.setSetpoint(0.67);
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/commands/ShooterCommand.java:31]: PID is unnecessary call subsystem.set() directly instead

File: src/main/java/frc/robot/commands/ShooterCommand.java [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+    SmartDashboard.putNumber("Shooter speed:", m_shooterSubsystem.m_flywheelMotor.getMotorOutputPercent());
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/commands/ShooterCommand.java:32]: this call should be in the subsystem

File: src/main/java/frc/robot/commands/ShooterCommand.java [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+  @Override
+  public void end(boolean interrupted) {
+    m_shooterSubsystem.setFlywheelPower(0);
+  }
+
+  // Returns true when the command should end.
+  @Override
+  public boolean isFinished() {
+    return false;
+  }
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/commands/ShooterCommand.java:46]: remove

File: src/main/java/frc/robot/commands/ShooterCommand.java [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+
+  // Called every time the scheduler runs while the command is scheduled.
+  @Override
+  public void execute() { 
+    m_pidController.setSetpoint(0.67);
+    SmartDashboard.putNumber("Shooter speed:", m_shooterSubsystem.m_flywheelMotor.getMotorOutputPercent());
+    m_shooterSubsystem.setFlywheelPower(m_pidController.calculate(m_shooterSubsystem.getFlywheelPower()));
+  }
+
+  // Called once the command ends or is interrupted.
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/commands/ShooterCommand.java:36]: remove

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+
+package frc.robot.subsystems;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
+
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.HardwareMap.ShooterHardware;
+
+public class ShooterSubsystem extends SubsystemBase {
+  /** Creates a new shooterSubsystem. */
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java:13]: how did this happen

**Review by LennyAbbas — CHANGES_REQUESTED** [review]
this is the worst code I have ever seen. Overall it needs organization and documentation.

**Comment by LennyAbbas** [comment]
once all my changes are resolved please request my review again

File: src/main/java/frc/robot/HardwareMap.java [hunk 25]
```diff
@@ -70,12 +72,26 @@
...
 
+  /** File for storing the hardware of the shooter. */
+  public class ShooterHardware {
+    public WPI_TalonFX shooter = new WPI_TalonFX(Constants.ShooterConstants.kShooterMotorPort);
+  }
+
+  /** File for storing the hardware of the inputs. */
+  public class InputHardware {
+    public XboxController driveController = new XboxController(Constants.OIConstants.kDriverControllerPort);
+  }
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/HardwareMap.java:83]: input hardware should not be added on this PR

**Review by LennyAbbas — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/RobotContainer.java [hunk 26]
```diff
@@ -34,11 +36,12 @@
...
   private SwerveDriveSubsystem m_swerveDriveSubsystem = new SwerveDriveSubsystem(
       m_hardwareMap.swerveDrivetrainHardware);
-
+  private ShooterSubsystem m_shooterSubsystem = new ShooterSubsystem(m_hardwareMap.shooterHardware);
+  
   private final MoveCommand m_defaultMoveCommand;
   private final MoveCommand m_aimingMoveCommand;
 
-  private XboxController m_driveController = new XboxController(OIConstants.kDriverControllerPort);
+  private XboxController m_driveController = new XboxController(OIConstants.kDriverControllerPort);;
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/RobotContainer.java:44]: delete semi-colon

File: src/main/java/frc/robot/RobotContainer.java [hunk 27]
```diff
@@ -34,11 +36,12 @@
   private HardwareMap m_hardwareMap = new HardwareMap();
   private SwerveDriveSubsystem m_swerveDriveSubsystem = new SwerveDriveSubsystem(
       m_hardwareMap.swerveDrivetrainHardware);
-
+  private ShooterSubsystem m_shooterSubsystem = new ShooterSubsystem(m_hardwareMap.shooterHardware);
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/RobotContainer.java:39]: if this subsystem is only used once it should be constructed inline

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+    m_targetPower = 0;
+  }
+
+  public void setPower(double power){
+    m_targetPower = power;
+  }
+
+  public double getFlywheelRPM(){
+    return m_flywheelMotor.getSelectedSensorVelocity();
+  }
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java:33]: this is useless

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+
+  @Override
+  public void periodic() {
+    // This method will be called once per scheduler run
+    m_flywheelMotor.set(m_targetPower);
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java:38]: call this in set()

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+    SmartDashboard.putNumber("Power", m_targetPower);
+    SmartDashboard.putNumber("Target RPM", m_targetPower*6380);
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java:40]: you only need to print the target speed and current speed

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 31]
```diff
@@ -0,0 +1,44 @@
...
+    SmartDashboard.putNumber("Current RPM", getFlywheelRPM());
+  }
+
+}
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java:44]: formatting in this file is wonk

**Review by LennyAbbas — CHANGES_REQUESTED** [review]
formatting in hardware map is wonk

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 32]
```diff
@@ -13,28 +13,32 @@
...
+  }
+
+  public double getFlywheelRPM(){
+    return m_flywheelMotor.getSelectedSensorVelocity();
   }
 
   @Override
   public void periodic() {
-    SmartDashboard.putNumber("Shooter Speed", m_flywheelMotor.get());
+    // This method will be called once per scheduler run
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java:37]: where did this comment come from?

**Comment by LennyAbbas** [comment]
also add a line in the README that says the shooter works

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 33]
```diff
@@ -0,0 +1,82 @@
...
+   * at 0.67 4160 +/- 10
+   * at 0.5 3110 +/- 10
+   * at 0.33 1980 +/- 10
+   * switching to karthik at 33%
+   * at 1 5935 +/- 10
+   * at 0.8 4640 +/- 10
+   * at 0.67 3935 +/- 15
+   * at 0.5 2950 +/- 10
+   * at 0.33 1875 +/- 10
+   */
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java:52]: should be removed before merge

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 34]
```diff
@@ -0,0 +1,82 @@
...
+    SmartDashboard.putNumber("Current Shooter RPM", getFlywheelRPM());
+  }
+
+  /*
+   * I was too lazy to make this a java doc, so this is what you may need to know
+   * when making one
+   * 4096 ticks per rotation
+   * 60,000 ms in a min and m_shooterMotor.getSelectedSensorVelocity() gives
+   * ticks/100ms
+   */
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java:68]: should be changed to a comment where relevant

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 35]
```diff
@@ -0,0 +1,82 @@
...
+  public double getFlywheelRPM() {
+    return m_shooterMotor.getSelectedSensorVelocity() * 600 / 2048;
+  }
```
Comment (LennyAbbas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java:71]: should be removed and the coefficient set by calling a builtin method for the talon fx

**Review by LennyAbbas — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java (hunk 35 above)
Comment (ArnoUUU) [reply src/main/java/frc/robot/subsystems/ShooterSubsystem.java:71]: for some reason my commits wont sync

Comment (LennyAbbas) [reply src/main/java/frc/robot/subsystems/ShooterSubsystem.java:71]: What do you mean?

Comment (ArnoUUU) [reply src/main/java/frc/robot/subsystems/ShooterSubsystem.java:71]: when I click sync changes it shows the "syncing" thing forever

Comment (ArnoUUU) [reply src/main/java/frc/robot/subsystems/ShooterSubsystem.java:71]: pretty much my changes aren't being pushed, so I can't resolve these conversations

Comment (LennyAbbas) [reply src/main/java/frc/robot/subsystems/ShooterSubsystem.java:71]: If you're working from home it might be because you're not signed into github properly which if you want help with we can discuss on slack

**Review by ProgrammingSR — APPROVED** [review]

**Review by LennyAbbas — APPROVED** [review]
